In [ ]:
!pip install qiskit

In [ ]:
from qiskit import QuantumRegister, QuantumCircuit, ClassicalRegister
from qiskit.circuit import Parameter, Instruction, Gate
from qiskit.circuit.library import *
import numpy as np
import matplotlib.pyplot as plt


#[Single Qubit Gates](https://youtu.be/egbvHi6v08M?si=BymP2Bc8VQeIaslA)

In [ ]:
from qiskit.circuit.library import SXdgGate

sxdagger = SXGate()
print(sxdagger.to_matrix())

[[0.5+0.5j 0.5-0.5j]
 [0.5-0.5j 0.5+0.5j]]


# Two-Qubit and Controlled Gates in Qiskit
*In quantum computing, controlled means:*

*“Apply a gate only if another qubit has a specific value.”*

---

# CHGate

## Controlled Hadamard Gate

Applies H gate when control qubit is 1.

$$
CH =
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&\frac1{\sqrt2}&\frac1{\sqrt2}\\
0&0&\frac1{\sqrt2}&-\frac1{\sqrt2}
\end{bmatrix}
$$

```python
qc.ch(0,1)
```

---

# CPhaseGate

## Controlled Phase Gate

$$
CP(\lambda)=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&1&0\\
0&0&0&e^{i\lambda}
\end{bmatrix}
$$

```python
qc.cp(theta,0,1)
```

---

# CRXGate

## Controlled RX Rotation

$$
CR_X(\theta)=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&\cos\frac\theta2&-i\sin\frac\theta2\\
0&0&-i\sin\frac\theta2&\cos\frac\theta2
\end{bmatrix}
$$

```python
qc.crx(theta,0,1)
```

---

# CRYGate

## Controlled RY Rotation

$$
CR_Y(\theta)=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&\cos\frac\theta2&-\sin\frac\theta2\\
0&0&\sin\frac\theta2&\cos\frac\theta2
\end{bmatrix}
$$

```python
qc.cry(theta,0,1)
```

---

# CRZGate

## Controlled RZ Rotation

$$
CR_Z(\theta)=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&e^{-i\theta/2}&0\\
0&0&0&e^{i\theta/2}
\end{bmatrix}
$$

```python
qc.crz(theta,0,1)
```

---

# CSGate

## Controlled S Gate

$$
CS=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&1&0\\
0&0&0&i
\end{bmatrix}
$$

```python
qc.cs(0,1)
```

---

# CSdgGate

## Controlled S-Dagger Gate

$$
CS^\dagger=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&1&0\\
0&0&0&-i
\end{bmatrix}
$$

```python
qc.csdg(0,1)
```

---

# CSXGate

## Controlled √X Gate

Applies SX when control = 1.

$$
(SX)^2 = X
$$

```python
qc.csx(0,1)
```

---

# CUGate

## Controlled U Gate

General controlled single-qubit unitary.

$$
CU(\theta,\phi,\lambda,\gamma)
$$

```python
qc.cu(theta,phi,lambda_,gamma,0,1)
```

---

# CU1Gate

## Controlled U1 Gate

Controlled phase rotation.

```python
qc.cu1(theta,0,1)
```

---

# CU3Gate

## Controlled U3 Gate

Controlled generic rotation.

```python
qc.cu3(theta,phi,lambda_,0,1)
```

---

# CXGate

## Controlled-X (CNOT)

Most important entangling gate.

$$
CX=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&0&1\\
0&0&1&0
\end{bmatrix}
$$

```python
qc.cx(0,1)
```

---

# CYGate

## Controlled-Y Gate

$$
CY=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&0&-i\\
0&0&i&0
\end{bmatrix}
$$

```python
qc.cy(0,1)
```

---

# CZGate

## Controlled-Z Gate

$$
CZ=
\begin{bmatrix}
1&0&0&0\\
0&1&0&0\\
0&0&1&0\\
0&0&0&-1
\end{bmatrix}
$$

```python
qc.cz(0,1)
```

---

# DCXGate

## Double-CNOT Gate

Equivalent to two CX gates.

$$
DCX = CX_{01}CX_{10}
$$

```python
qc.dcx(0,1)
```

---

# ECRGate

## Echoed Cross Resonance Gate

Hardware-native IBM gate.

Used for superconducting qubit interactions.

```python
qc.ecr(0,1)
```

---

# iSwapGate

## iSWAP Gate

Swaps qubits with phase factor \(i\).

$$
iSWAP=
\begin{bmatrix}
1&0&0&0\\
0&0&i&0\\
0&i&0&0\\
0&0&0&1
\end{bmatrix}
$$

```python
qc.iswap(0,1)
```

---

# RXXGate

## XX Interaction Rotation

$$
R_{XX}(\phi)=e^{-i\phi XX/2}
$$

Rotation around:

$$
X \otimes X
$$

```python
qc.rxx(phi,0,1)
```

---

# RYYGate

## YY Interaction Rotation

$$
R_{YY}(\phi)=e^{-i\phi YY/2}
$$

Rotation around:

$$
Y \otimes Y
$$

```python
qc.ryy(phi,0,1)
```

---

# RZXGate

## ZX Interaction Rotation

$$
R_{ZX}(\phi)=e^{-i\phi ZX/2}
$$

Rotation around:

$$
Z \otimes X
$$

```python
qc.rzx(phi,0,1)
```

---

# RZZGate

## ZZ Interaction Rotation

$$
R_{ZZ}(\phi)=e^{-i\phi ZZ/2}
$$

Rotation around:

$$
Z \otimes Z
$$

```python
qc.rzz(phi,0,1)
```

---

# SwapGate

## SWAP Gate

Swaps two qubits.

$$
SWAP=
\begin{bmatrix}
1&0&0&0\\
0&0&1&0\\
0&1&0&0\\
0&0&0&1
\end{bmatrix}
$$

```python
qc.swap(0,1)
```

---

# XXMinusYYGate

## XX-YY Interaction Gate

Parameterized interaction gate:

$$
XX - YY
$$

Used in quantum simulations.

```python
qc.xx_minus_yy(theta,beta,0,1)
```

---

# XXPlusYYGate

## XX+YY Interaction Gate

Parameterized interaction gate:

$$
XX + YY
$$

Useful in fermionic simulations.

```python
qc.xx_plus_yy(theta,beta,0,1)
```

---

# Summary

| Gate | Purpose |
|---|---|
| CX | Controlled-X |
| CZ | Controlled phase flip |
| SWAP | Swap qubits |
| iSWAP | Swap with phase |
| CRX/CRY/CRZ | Controlled rotations |
| RXX/RYY/RZZ | Interaction rotations |
| ECR | Hardware-native interaction |
| CH | Controlled Hadamard |
| CU | General controlled unitary |

In [ ]:
from qiskit.circuit.library import CHGate

chgate = CHGate()

print(CHGate().to_matrix())

[[ 1.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j]
 [ 0.        +0.j  0.70710678+0.j  0.        +0.j  0.70710678+0.j]
 [ 0.        +0.j  0.        +0.j  1.        +0.j  0.        +0.j]
 [ 0.        +0.j  0.70710678+0.j  0.        +0.j -0.70710678+0.j]]


In [ ]:
from qiskit.circuit.library import SwapGate

swapgate = SwapGate()

print(SwapGate().to_matrix())

[[1.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 1.+0.j]]


# Three-Qubit and Multi-Qubit Standard Gates in Qiskit

---

# C3SXGate

## Triple-Controlled √X Gate

Applies the √X gate when all 3 control qubits are 1.

Condition:

$$
|111\rangle \rightarrow SX
$$

Property:

$$
(SX)^2 = X
$$

```python
qc.c3sx(c1,c2,c3,target)
```

---

# C3XGate

## Triple-Controlled X Gate

Applies X only if all 3 controls are 1.

Also called a 3-controlled Toffoli gate.

Logical action:

$$
|1110\rangle \rightarrow |1111\rangle
$$

```python
qc.c3x(c1,c2,c3,target)
```

---

# C4XGate

## Four-Controlled X Gate

Applies X only if all 4 control qubits are 1.

Logical action:

$$
|11110\rangle \rightarrow |11111\rangle
$$

```python
qc.c4x(c1,c2,c3,c4,target)
```

---

# CCXGate

## Toffoli Gate

Controlled-Controlled-X gate.

Two controls + one target.

Rule:

```text
If both controls = 1:
    apply X
```

---

## Matrix

$$
CCX=
\begin{bmatrix}
1&0&0&0&0&0&0&0\\
0&1&0&0&0&0&0&0\\
0&0&1&0&0&0&0&0\\
0&0&0&1&0&0&0&0\\
0&0&0&0&1&0&0&0\\
0&0&0&0&0&1&0&0\\
0&0&0&0&0&0&0&1\\
0&0&0&0&0&0&1&0
\end{bmatrix}
$$

Only the last two states swap:

$$
|110\rangle \leftrightarrow |111\rangle
$$

```python
qc.ccx(0,1,2)
```

---

# CCZGate

## Controlled-Controlled-Z Gate

Applies Z phase flip when both controls are 1.

---

## Matrix

$$
CCZ=
\begin{bmatrix}
1&0&0&0&0&0&0&0\\
0&1&0&0&0&0&0&0\\
0&0&1&0&0&0&0&0\\
0&0&0&1&0&0&0&0\\
0&0&0&0&1&0&0&0\\
0&0&0&0&0&1&0&0\\
0&0&0&0&0&0&1&0\\
0&0&0&0&0&0&0&-1
\end{bmatrix}
$$

The state:

$$
|111\rangle
$$

gets a phase:

$$
-1
$$

```python
qc.ccz(0,1,2)
```

---

# CSwapGate

## Controlled SWAP Gate (Fredkin Gate)

Swaps two target qubits if control qubit = 1.

---

## Action

$$
|1ab\rangle \rightarrow |1ba\rangle
$$

if control is 1.

```python
qc.cswap(control,target1,target2)
```

---

# RCCXGate

## Relative-Phase Toffoli Gate

Simplified Toffoli gate.

Uses fewer quantum operations.

More hardware efficient.

Not exactly equal to CCX because relative phases differ.

---

## Key Idea

Logical computation is preserved, but some phases change.

```python
qc.rccx(0,1,2)
```

---

# RC3XGate

## Relative-Phase 3-Controlled X Gate

Simplified version of C3X.

Uses fewer gates and lower circuit depth.

More efficient for real hardware.

```python
qc.rc3x(c1,c2,c3,target)
```

---

# Important Concept

## Standard Toffoli

Exact logical operation.

Higher gate cost.

---

## Relative-Phase Versions

Approximate phase behavior:

- cheaper
- faster
- fewer CX gates
- useful in optimization

---

# Matrix Size Rule

For \(n\) qubits:

$$
2^n \times 2^n
$$

Examples:

| Qubits | Matrix Size |
|---|---|
| 1 | \(2 \times 2\) |
| 2 | \(4 \times 4\) |
| 3 | \(8 \times 8\) |
| 4 | \(16 \times 16\) |

---

# Summary Table

| Gate | Meaning |
|---|---|
| CCX | Toffoli gate |
| CCZ | Controlled-Controlled-Z |
| CSwap | Fredkin gate |
| C3X | Triple-controlled X |
| C4X | Four-controlled X |
| C3SX | Triple-controlled √X |
| RCCX | Simplified Toffoli |
| RC3X | Simplified 3-control Toffoli |

---

# Core Idea

Multi-controlled gates apply operations only when:

$$
\text{all control qubits} = |1\rangle
$$

These gates are essential for:

- reversible computing
- arithmetic circuits
- Grover's algorithm
- quantum error correction
- logical control operations

#[Toffoli **/** CCX gate](https://youtu.be/78Kfu5XIV3E?si=PZXSLF6F9JrtXsh_)

In [ ]:
from qiskit.circuit.library import CCXGate

ccxgate = CCXGate()

print(ccxgate.to_matrix())

[[1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]]


# Global Phase in Qiskit

---

# Definition

A global phase multiplies the entire quantum state by:

$$
e^{i\phi}
$$

where:

- $i = \sqrt{-1}$
- $\phi$ = phase angle

---

# Quantum State

Original state:

$$
|\psi\rangle
$$

After global phase:

$$
e^{i\phi}|\psi\rangle
$$

---

# Important Property

Global phase does NOT change measurement probabilities.

Because:

$$
|e^{i\phi}| = 1
$$

Probabilities use:

$$
|a|^2
$$

so the phase disappears during measurement.

---

# Example

State:

$$
|\psi\rangle=
\begin{bmatrix}
0.6 \\
0.8
\end{bmatrix}
$$

Apply global phase:

$$
i|\psi\rangle=
\begin{bmatrix}
0.6i \\
0.8i
\end{bmatrix}
$$

Probabilities remain:

$$
|0.6i|^2 = 0.36
$$

$$
|0.8i|^2 = 0.64
$$

Total:

$$
0.36 + 0.64 = 1
$$

---

# Global Phase Operator

$$
e^{i\phi}I
$$

Where:

- $I$ = identity matrix

---

# Important Difference

## Global Phase

Applies equally to the whole state:

$$
e^{i\phi}|\psi\rangle
$$

Usually not observable.

---

## Relative Phase

Changes amplitudes differently.

Example:

$$
|0\rangle + |1\rangle
$$

vs

$$
|0\rangle - |1\rangle
$$

Relative phase affects interference and computation.

---

# In Qiskit

Set global phase:

```python
qc.global_phase = 3.14
```

or:

```python
QuantumCircuit(2, global_phase=3.14)
```

---

# Why Qiskit Tracks It

Global phase is important for:

- circuit equivalence
- transpilation
- decomposition
- unitary calculations

Two circuits may differ only by global phase but still behave identically physically.

---

# Core Idea

Global phase rotates the entire quantum state equally:

$$
e^{i\phi}
$$

It usually has no measurable physical effect, but is mathematically important in quantum computing.

In [ ]:
from qiskit.circuit.library import GlobalPhaseGate

globalP = GlobalPhaseGate(np.pi/2)

print(globalP.to_matrix())

[[6.123234e-17+1.j]]


# GlobalPhaseGate Example in Qiskit

---

# Code Example

```python
from qiskit.circuit.library import GlobalPhaseGate
import numpy as np

gpgate = GlobalPhaseGate(np.pi/2)

print(gpgate.to_matrix())
```

---

# Output

```python
[[6.123234e-17+1.j]]
```

---

# Why This Happens

The gate applies a global phase:

$$
e^{i\phi}
$$

For:

$$
\phi = \frac{\pi}{2}
$$

we get:

$$
e^{i\pi/2}
$$

---

# Euler Formula

$$
e^{i\theta}
=
\cos(\theta)+i\sin(\theta)
$$

Substitute:

$$
\theta = \frac{\pi}{2}
$$

Then:

$$
e^{i\pi/2}
=
\cos\left(\frac{\pi}{2}\right)
+
i\sin\left(\frac{\pi}{2}\right)
$$

Since:

$$
\cos\left(\frac{\pi}{2}\right)\approx 0
$$

and

$$
\sin\left(\frac{\pi}{2}\right)=1
$$

we get:

$$
e^{i\pi/2}=i
$$

---

# Why Tiny Number Appears

Computer calculations use floating-point approximations.

Instead of exact zero:

$$
0
$$

Python shows:

```python
6.123234e-17
```

which means:

$$
6.123234 \times 10^{-17}
$$

This is extremely close to zero.

So:

```python
6.123234e-17 + 1.j
```

really means:

$$
0 + i
$$

or simply:

$$
i
$$

---

# Why Matrix Size is 1×1

Global phase is just a scalar multiplier:

$$
e^{i\phi}
$$

It does not act on specific qubits like X or CX gates.

So Qiskit may represent it internally as:

$$
[[i]]
$$

instead of a larger identity matrix.

---

# Important Concept

Global phase multiplies the entire quantum state equally:

$$
|\psi\rangle
\rightarrow
e^{i\phi}|\psi\rangle
$$

It usually does NOT change measurable probabilities.

---

# Core Idea

Your output:

```python
[[6.123234e-17+1.j]]
```

is mathematically equal to:

$$
i
$$

which is the correct global phase for:

$$
\phi=\frac{\pi}{2}
$$